classpath/java/io/BufferedInputStream.java

```diff
@@ -54,28 +59,7 @@ public int read(byte[] b, int offset, int length) throws IOException {
 
       count += remaining;
       position += remaining;
-      offset += remaining;
-      length -= remaining;
     }
-
-    while (length > 0) {
```

Comment:
As you say... The spec says: "This method blocks until input data is available"! And there IS data available in the buffer. If the buffer is empty the method may still block while filling the buffer. 
Also the current method does not even try to fill the buffer if "read(byte[] b, int offset, int length)" is called. This is very bad if the given buffer "b" is small. This could end up in reading single bytes from the network, if b is of size 1, which would clearly make the BufferedInputStream useless.